src/java.base/share/classes/java/util/Collection.java

```diff
@@ -740,6 +740,7 @@ default boolean removeIf(Predicate<? super E> filter) {
      * @since 1.8
      */
     default Stream<E> stream() {
+        if (isEmpty()) return Stream.empty();
```

Comment:
The net effect of this change might depend on your workload. If you call stream() on empty collections that have cheap isEmpty(), this change will likely improve performance and reduce waste. However, this same change might do the opposite if some of your collections aren't empty or have costly isEmpty(). Concurrent collections come to mind. It would be good to have benchmarks for different workloads.

Reply:
wouldn't this make streams no longer lazy if the collection is empty?
```java
        List<String> list = new ArrayList<>();
        Stream<String> stream = list.stream();

        list.addAll(List.of("one", "two", "three"));

        stream.forEach(System.out::println); // prints one two three
```

Reply:
(immutable collections could override stream() instead, since they don't have that problem)

Reply:
> The net effect of this change might depend on your workload. If you call stream() on empty collections that have cheap isEmpty(), this change will likely improve performance and reduce waste. However, this same change might do the opposite if some of your collections aren't empty or have costly isEmpty(). It would be good to have benchmarks for different workloads.

Yes, I also thought about the cost of isEmpty() on concurrent collections. There are four concurrent collections that have a linear time cost size() method: CLQ, CLD, LTQ and CHM. However, in each of these cases, the isEmpty() method has constant time cost. There might be collections defined outside the JDK where this could be the case.

However, I will extend the benchmark to include a few of those cases too, as well as different sizes and collection sizes.

Thank you so much for your input.

Reply:
> wouldn't this make streams no longer lazy if the collection is empty?
> 
> ```java
>         List<String> list = new ArrayList<>();
>         Stream<String> stream = list.stream();
> 
>         list.addAll(List.of("one", "two", "three"));
> 
>         stream.forEach(System.out::println); // prints one two three
> ```

I did not consider this case, thank you for bringing it up. I have always found this behaviour a bit strange and have never used it "in the real world". It is also not consistent between collections. Here is an example with four collections: ArrayList, CopyOnWriteArrayList, ConcurrentSkipListSet and ArrayBlockingQueue:

```java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.ConcurrentSkipListSet;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Supplier;
import java.util.stream.IntStream;

public class LazyStreamDemo {
    public static void main(String... args) {
        List<Supplier<Collection<String>>> suppliers =
                List.of(ArrayList::new, // fast-fail
                        CopyOnWriteArrayList::new, // snapshot
                        ConcurrentSkipListSet::new, // weakly-consistent
                        () -> new ArrayBlockingQueue<>(10) // weakly-consistent
                );
        for (Supplier<Collection<String>> supplier : suppliers) {
            Collection<String> c = supplier.get();
            System.out.println(c.getClass());
            IntStream stream = c.stream()
                    .sorted()
                    .filter(Objects::nonNull)
                    .mapToInt(String::length)
                    .sorted();

            c.addAll(List.of("one", "two", "three", "four", "five"));
            System.out.println("stream = " + Arrays.toString(stream.toArray()));
        }
    }
}
```

The output is:

```
class java.util.ArrayList
stream = [3, 3, 4, 4, 5]
class java.util.concurrent.CopyOnWriteArrayList
stream = []
class java.util.concurrent.ConcurrentSkipListSet
stream = []
class java.util.concurrent.ArrayBlockingQueue
stream = [3, 3, 4, 4, 5]
```

At least with the EmptyStream we would have consistent output of always []

Reply:
@kabutz I agree that i wouldn't consider it clean code to use a stream like i put into the example. I only brought it up because it might break existing code, since i think streams are expected to be lazy. Interesting to see that they are in fact not lazy in all situations - i put that into my notes.

Reply:
The Javadoc for [java.util.stream](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/stream/package-summary.html) is clear though:

> Traversal of the pipeline source does not begin until the terminal operation of the pipeline is executed.

and further on says:

> Spliterators for mutable data sources have an additional challenge; timing of binding to the data, since the data could change between the time the spliterator is created and the time the stream pipeline is executed. Ideally, a spliterator for a stream would report a characteristic of IMMUTABLE or CONCURRENT; if not it should be late-binding.

which explains why the collections in `java.util.concurrent` (whose spliterators have the `CONCURRENT` characteristic) need not be late-binding.

So this change is backward incompatible and will break code in subtle ways (@mbien's code is a minimal example, but in real-world code, it may be much less obvious that the collection may be empty on stream creation and that the collection is modified between stream creation and pipeline execution).
Moreover, in cases where this is indeed a performance bottleneck, it's easy enough for the developer to short-circuit the pipeline themselves in case of an empty source.

Reply:
Thanks @anthonyvdotbe I believe that would satisfy the lazy binding, and then we should increase the use of the IMMUTABLE characteristic where it makes sense.